### DISCUSSION TITLE
Nebula Graph 索引详解 - siwei.io

### DISCUSSION THREAD

**Discussion by giscus** [siwei-io/siwei-io.github.io#10] [Comments]
# Nebula Graph 索引详解 - siwei.io

找不到索引？为什么我要创建 Nebula Graph 索引？什么时候要用到 Nebula Graph 原生索引，一文把这些搞清楚。

https://www.siwei.io/nebula-index-explained/

**Comment by abby-cyber** [comment]
Bravo!
How can I give a like to this post?

**Reply by wey-gu** [reply]
Thanks a lot, @abby-cyber , this means a lot to me, btw, you could actually do reactions from above:

<img width="1220" alt="Screen Shot 2022-08-04 at 2 54 00 PM" src="https://user-images.githubusercontent.com/1651790/182782897-7d76583e-009b-48c0-bde3-5bbbd0c75911.png">

**Reply by abby-cyber** [reply]
Just did it!

**Comment by wey-gu** [comment]
更新，现在对于无索引的情况下，我们支持这样的表达:

点的例子
```sql
MATCH (v) WITH properties(v) AS props, keys(properties(v)) AS kk, v LIMIT 10000 WHERE [i IN kk WHERE props[i] == "Tony Parker"] RETURN v
```

边的例子
```sql
MATCH ()-[e]->() WITH e, properties(e) AS props, keys(properties(e)) AS kk LIMIT 10000 WHERE [i IN kk WHERE props[i] == 90] RETURN e
```

要点

- LIMIT 之后可以 WHERE ！！！！
- 现在支持无类型的 vertex/edge sample scan
- 善用 WHERE 中的列表展开

ref: https://discuss.nebula-graph.com.cn/t/topic/9936?u=wey by @jievince

**Comment by abby-cyber** [comment]
请教下两个问题：
1. 基于规则的优化选择索引，Fan Out 到所有 StorageD  这句话咋理解呀？
2. 经常听到下推，什么情况下可以下推？这个是我们代码写死了可以下推么，还是说需要根据过滤条件决定可不可以下推呢？

**Reply by wey-gu** [reply]
1. 基于规则的优化选择索引，Fan Out 到所有 StorageD 这句话咋理解呀？

> 规则就是不是基于 cost 的(比如估计怎么优化的代价更大，选择更小的那一个分支），相反所有选择都是基于预制的模式匹配判断，这个理解吧哈？
fan out 是说广播到所有 storageD 这里指的是，索引和数据在一起，不是 unified 的索引，所以查询数据需要广播到space 的所有 storaged，这是非统一索引所带来的后果。
> 具体来说，如果你要 limit 100，如果你落在100个 storaged 上，其实你还是取了 最多 100*100 条，在limit 下推的情况下依然如此


2. 经常听到下推，什么情况下可以下推？这个是我们代码写死了可以下推么，还是说需要根据过滤条件决定可不可以下推呢？

> 下推就是比如 filter（follow.degree>50) 下推，的话，getNeibrs 就只扫描符合条件的给 graphd，而不下推则是扫完了，在graphd 去 filter。另外，这里也有介绍哈  https://www.siwei.io/ngql-execution-plan/

感谢 @abby-cyber 提出的好问题（还辛苦把 IM 里的问题又搬运过来）
